usr.bin/whereis/whereis.1
.\" Copyright (c) 1980 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)whereis.1	5.1 (Berkeley) 4/29/85
.\"
.TH WHEREIS 1 "24 February 1979"
.UC
.SH NAME
whereis \- locate source, binary, and or manual for program
.SH SYNOPSIS
.B whereis
[
.B \-sbm
] [
.B \-u
] [
.B \-SBM
dir ...
.B \-f
] name ...
.SH DESCRIPTION
.I Whereis
locates source/binary and manuals sections for specified files.
The supplied names are first stripped of leading pathname components
and any (single) trailing extension of the form ``.ext'', e.g. ``.c''.
Prefixes of ``s.'' resulting from use of source code control are also
dealt with.
.I Whereis
then attempts to locate the desired program in a list of standard places.
If any of the
.B \-b,
.B \-s
or
.B \-m
flags are given then
.I whereis
searches only for binaries, sources or manual sections respectively
(or any two thereof).
The
.B \-u
flag may be used to search for unusual entries.
A file is said to be unusual if it does not have one entry of
each requested type.
Thus ``whereis -m -u *'' asks for those files in the current
directory which have no documentation.
.sp
Finally, the
.B \-B
.B \-M
and
.B \-S
flags may be used to change or otherwise limit the places where
.I whereis
searches.
The
.B \-f
file flags is used to terminate the last such directory list
and signal the start of file names.
.SH EXAMPLE
The following finds all the files in /usr/bin which are not documented
in /usr/man/man1 with source in /usr/src/cmd:
.IP
cd /usr/ucb
.br
whereis \-u \-M /usr/man/man1 \-S /usr/src/cmd \-f *
.SH FILES
/usr/src/*
.br
/usr/{doc,man}/*
.br
/lib, /etc, /usr/{lib,bin,ucb,old,new,local}
.SH AUTHOR
William Joy
.SH BUGS
Since the program uses
.IR chdir (2)
to run faster, pathnames given with the
.B \-M
.B \-S
and
.B \-B
must be full; i.e. they must begin with a ``/''.


ChangeLog
===========================================================================
2020-10-20
Tuesday, October 20, 2020

Started a ChangeLog file to keep track of my daily progress. I will need to
add in issues I have previously come up with and that which are not yet
included here.

I pushed a commit today of GPIO toggling (for the onboard LEDs) which goes
into great detail of each required part.
There are four registers to configure, in addition to configuring the RCC
clock. There are two ways to set or reset a GPIO pin once it is all set up:
by a read/modify/write dance, or with an atomic bit set or reset operation.

For setting up the GPIO (for port D, for the LEDs), this needs to be done:
 Set up the gpio:
1. clock through the RCC_AHB1ENR register.
2. mode through the GPIOD_MODER register.
3. output as push-pull or open-drain through the GPIOD_OTYPER register.
4. port speed through the GPIOD_OSPEEDR register.
5. pull-up or pull-down resistors through the GPIOD_PUPDR register.

For actually using the GPIO as an output there are two ways:
1. Read/modify/write of the GPIOD_ODR Output Data Register
2. Writing to the GPIOD_BSRR Bit Set/Reset Register

Whichever one is used (or available) is chip vendor dependent. These are
both available for STM32F4 chips.

I put the linker script in the testing base directory (with Makefile.inc)
so that all testing programs can use it, instead of having each program
duplicating the linker script.

The program I finished today has links to all the documentation I have used
to make that program. I will end up putting them in this repository.

I added a .gitignore to the testing directory. Not sure if it is best to
use a global one yet, so just this local one for now.

===========================================================================
2020-10-31
Saturday, October 31, 2020

Pushed two commits that import the RetroBSD repo into this project repo.
I also included the zip from the autobuilder so that there is some starting
point that is easily referenced (instead of just a commit id).


The last while I've been trying to get a mips-elf port of gcc, binutils,
and gdb to build the RetroBSD repo (which is mips-based). I had already got
it to work on Ubuntu (through the Windows Subsystem for Linux) by following
the instructions at http://retrobsd.org/wiki/doku.php/doc/toolchain-mips
which took quite a bit of time to go through.

Also, these instructions are made for an old version of gcc, compiled with
another old version of gcc. The newer gcc's are way more pedantic and find
so many more bugs, even small ones, so it has been a challenge. For
instance, one bug I came across was that glibc changed the location of some
defines into a different header, so I had to find out which one it was and
make a patch for it. It's not hard, just tedious and unfulfilling work.


Get first crack at a devel/mips-elf OpenBSD port done. It has gcc 4.8.1 and
binutils 2.27. Many patches were made. 8 for binutils, 19 for gcc.
I based this port on the ports devel/xtensa-elf and devel/arm-none-eabi.
gdb and newlib are not done yet (and maybe never will be).
binutils was the first I got working. It took a while to get gcc to work.

This port is for compiling the original code of RetroBSD on OpenBSD. The
original code targets the PIC32 MIPS-based microcontroller, thus mips-elf.

There are two or three segfaults in the process of building gcc. I have no
reason why, and I haven't investigated. Starting the build over again
allows a bit more progress until there is another segfault. I just kept
restarting it until the build is complete. The segfaults could be due to
running out of memory or something like that. Who knows. gcc is complex.

There are some post-patch and post-install fixes in the devel/avr/gcc
Makefile that need to be used. This will get rid of all the ridiculous
patches for the single space character needed by c++11 for variables.

===========================================================================
2020-11-01
Sunday, November 01, 2020

I started to build DiscoBSD with my new toolchain, but there are quite a
few errors. The main one is (after much investigation into the cause) due
to gcc not passing -msoft-float to the assembler. This caused the assembler
to assemble as -mhard-float. The linker wasn't able to link these two
separate types of objects.

I fixed this by getting a patchset from gcc, which I had to modify, as it
was meant for some newer version of gcc (5?).

I made 4 new patches, and added some more changes to an already existing
patch. That fixed the whole floating-point issue. The userland programs can
be built now. The kernel builds, too. There still is a problem with the
linker overlaying sections of the executable. I haven't researched it yet,
though.

The VirtualMIPS simulator isn't building. Probably due to not having the
libelf package on OpenBSD. I may need to port that library, too. Who knows.

Other problems are: elf2aout, manpages, init_main.c, expr, yacc kconfig
reduce/reduce conflicts, install not working.

I updated the README.md file for tools/openbsd to include how to get
binutils and gcc for the MIPS target.


Now I am trying to get VirtualMIPS to work. Most errors are about specific
types used. I am guessing they are defined in the libelf code.

I have had to (temporarily) hard-code some directory paths in the Makefile
for VirtualMIPS to what OpenBSD uses, such as /usr/local/include instead of
/opt/local/include on Linux.


Some cool things about the name DiscoBSD:
First, 'disco' is Latin for 'I learn' or 'to learn' or 'become aquainted
with' or 'I study, practice'. All of these definitions are quite in line
with the whole purpose of this directed study, which is to do a heck of a
lot of learning about Arm and about Unix and about early BSD.

Second, STM has a 'Discovery' line of demonstration and evauation circuit
boards with STM32F4 microcontrollers. I am targeting the STM32F4DISCOVERY

usr.bin/whereis/whereis.c
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */
#include <sys/param.h>
#include <sys/dir.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

void    getlist(int *, char ***, char ***, int *);
void    zerof();
void    lookup(char *);
void    looksrc(char *);
void    lookbin(char *);
void    lookman(char *);
void    findv(char **, int, char *);
void    find(char **, char *);
void    findin(char *, char *);
int     itsit(char *, char *);

static char *bindirs[] = {
    "/etc",
    "/bin",
    "/sbin",
    "/usr/bin",
    "/usr/sbin",
    "/usr/games",
    "/usr/libexec",
    "/usr/local",
    "/usr/local/bin",
    "/usr/local/sbin",
    0
};
static char *mandirs[] = {
    "/usr/share/man/man1",
    "/usr/share/man/man2",
    "/usr/share/man/man3",
    "/usr/share/man/man4",
    "/usr/share/man/man5",
    "/usr/share/man/man6",
    "/usr/share/man/man7",
    "/usr/share/man/man8",
    "/usr/share/man/man9",
    "/usr/share/man/manl",
    "/usr/share/man/mann",
    "/usr/share/man/mano",
    0
};
static char *srcdirs[]  = {
    "/usr/src/bin",
    "/usr/src/sbin",
    "/usr/src/usr.bin",
    "/usr/src/usr.sbin",
    "/usr/src/etc",
    "/usr/src/games",
    "/usr/src/lib",
    "/usr/src/libexec",
    "/usr/src/include",
    "/usr/src/lib/libc/gen",
    "/usr/src/lib/libc/stdio",
    "/usr/src/lib/libc/sys",
    "/usr/src/lib/libc/inet",
    "/usr/src/lib/libc/net",
    "/usr/src/lib/libc/ns",
    0
};

char    sflag = 1;
char    bflag = 1;
char    mflag = 1;
char    **Sflag;
int Scnt;
char    **Bflag;
int Bcnt;
char    **Mflag;
int Mcnt;
char    uflag;

/*
 * whereis name
 * look for source, documentation and binaries
 */
int
main(argc, argv)
    int argc;
    char *argv[];
{
    argc--, argv++;
    if (argc == 0) {
usage:
        fprintf(stderr, "whereis [ -bms ] [ -u ] [ -BMS dir ... -f ] name ...\n");
        exit(1);
    }
    do
        if (argv[0][0] == '-') {
            register char *cp = argv[0] + 1;
            while (*cp) switch (*cp++) {

            case 'f':
                break;

            case 'S':
                getlist(&argc, &argv, &Sflag, &Scnt);
                break;

            case 'B':
                getlist(&argc, &argv, &Bflag, &Bcnt);
                break;

            case 'M':
                getlist(&argc, &argv, &Mflag, &Mcnt);
                break;

            case 's':
                zerof();
                sflag++;
                continue;

            case 'u':
                uflag++;
                continue;

            case 'b':
                zerof();
                bflag++;
                continue;

            case 'm':
                zerof();
                mflag++;
                continue;

            default:
                goto usage;
            }
            argv++;
        } else
            lookup(*argv++);
    while (--argc > 0);
}

void
getlist(argcp, argvp, flagp, cntp)
    char ***argvp;
    int *argcp;
    char ***flagp;
    int *cntp;
{
    (*argvp)++;
    *flagp = *argvp;
    *cntp = 0;
    for ((*argcp)--; *argcp > 0 && (*argvp)[0][0] != '-'; (*argcp)--)
        (*cntp)++, (*argvp)++;
    (*argcp)++;
    (*argvp)--;
}

void
zerof()
{
    if (sflag && bflag && mflag)
        sflag = bflag = mflag = 0;
}
int count;
int print;

void
lookup(cp)
    register char *cp;
{
    register char *dp;

    for (dp = cp; *dp; dp++)
        continue;
    for (; dp > cp; dp--) {
        if (*dp == '.') {
            *dp = 0;
            break;
        }
    }
    for (dp = cp; *dp; dp++)
        if (*dp == '/')
            cp = dp + 1;
    if (uflag) {
        print = 0;
        count = 0;
    } else
        print = 1;
again:
    if (print)
        printf("%s:", cp);
    if (sflag) {
        looksrc(cp);
        if (uflag && print == 0 && count != 1) {
            print = 1;
            goto again;
        }
    }
    count = 0;
    if (bflag) {
        lookbin(cp);
        if (uflag && print == 0 && count != 1) {
            print = 1;
            goto again;
        }
    }
    count = 0;
    if (mflag) {
        lookman(cp);
        if (uflag && print == 0 && count != 1) {
            print = 1;
            goto again;
        }
    }
    if (print)
        printf("\n");
}

void
looksrc(cp)
    char *cp;
{
    if (Sflag == 0) {
        find(srcdirs, cp);
    } else
        findv(Sflag, Scnt, cp);
}

void
lookbin(cp)
    char *cp;
{
    if (Bflag == 0)
        find(bindirs, cp);
    else
        findv(Bflag, Bcnt, cp);
}

void
lookman(cp)
    char *cp;
{
    if (Mflag == 0) {
        find(mandirs, cp);
    } else
        findv(Mflag, Mcnt, cp);
}

void
findv(dirv, dirc, cp)
    char **dirv;
    int dirc;
    char *cp;
{
    while (dirc > 0)
        findin(*dirv++, cp), dirc--;
}

void
find(dirs, cp)
    char **dirs;
    char *cp;
{
    while (*dirs)
        findin(*dirs++, cp);
}

void
findin(dir, cp)
    char *dir, *cp;
{
    DIR *dirp;
    struct direct *dp;

    dirp = opendir(dir);
    if (dirp == NULL)
        return;
    while ((dp = readdir(dirp)) != NULL) {
        if (itsit(cp, dp->d_name)) {
            count++;
            if (print)
                printf(" %s/%s", dir, dp->d_name);
        }
    }
    closedir(dirp);
}

int
itsit(cp, dp)
    register char *cp, *dp;
{
    register int i = strlen(dp);

    if (dp[0] == 's' && dp[1] == '.' && itsit(cp, dp+2))
        return (1);
    while (*cp && *dp && *cp == *dp)
        cp++, dp++, i--;
    if (*cp == 0 && *dp == 0)
        return (1);
    while (isdigit(*dp))
        dp++;
    if (*cp == 0 && *dp++ == '.') {
        --i;
        while (i > 0 && *dp)
            if (--i, *dp++ == '.')
                return (*dp++ == 'C' && *dp++ == 0);
        return (1);
    }
    return (0);
}
